File: Javascript/39. Combination Sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let result = []
    let temp = []

    function backtracking(left, res) {
        // console.log(temp)
        if (res < 0) return false;
        if (res == 0) {
            result.push([...temp])
        }

        for (let i = left; i < candidates.length; i++) {
            temp.push(candidates[i])
            backtracking(i+1, res - candidates[i])
            temp.pop()
        }
    }

    backtracking(0, target)
    return result
};

File: Python/39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        result = []
        temp = []
        def bactracking(arr, target, temp):
            for i in range(len(arr)):
                div = target // arr[i]

                if div == 0 or len(arr) == 0:
                    continue

                for j in range (1, div + 1):
                    if arr[i] * j == target:
                        result.append(temp + [arr[i]]*j)
                        continue
                    
                    bactracking(arr[:i], target - arr[i]*j, temp + [arr[i]]*j)
        bactracking(candidates, target, temp)
        return result

if __name__ == "__main__":
    solution = Solution()
    # For testing
    print("result:", solution.combinationSum([2,3,5], 11))